### DISCUSSION TITLE
Fun with Smart Contract Lottery...

### DISCUSSION THREAD

**Discussion by toddabraham** [smartcontractkit/full-blockchain-solidity-course-py#79] [Q&A]
Currently running a lottery unit test on `def test_get_entrance_fee` and noticed, when checking the conversion of USD$50 to eth, it's currently 0.016 instead of 0.025 as per the recording.  The test of course failed when I tried to replace `expected_entrance_fee = Web3.toWei(0.025, "ether")` with `expected_entrance_fee = Web3.toWei(0.016, "ether")`.  My problem is, I wasn't able to fully trace back to where in the chain of code, between the Lottery.sol file and supporting script files, should I make the change.  

It's of course, not really necessary for me to do this in order to continue on with the course, but I thought it a good exercise to test my understanding.

**Comment by AlbertoSinigaglia** [comment]
can you please be more clear on what you are missing? 

Your test is failing since the oracle is returning a different conversion from the one provided in the tutorial, since Ether value fluctuates... this means, that you can't actually test this with just a simple asserting, you need to get the conversion from the oracle in your python script, and then check with an assertion:
```python
dollars = 50
contract_result = contract.getConversionRate() # or whatever method is returning you the entrance fee
c = interface.IAggregatorV3("address of the aggregator in the network you are using for eth to usd")
(
    roundId,
    answer,
    startedAt,
    updatedAt,
    answeredInRound
) = c.latestRoundData()
# x = answer / (10 ** c.decimals())  ---  will give you 25XX.XX$ per eth
# y = 1 / x  ---  will give you 0.00NNN eth per $
# z = y * dollars  ---  will return the number of eth for each dollar
# Web3.toWei(z, "ether")   ---   will return the number of wei you should be paying

supposed_result = Web3.toWei((1 / answer / (10 ** c.decimals())) * dollars, "ether")
assert contract_result == supposed_result
```

However, It's very not recommended checking for equality when dealing with floating point numbers... check [this StackOverflow page](https://stackoverflow.com/questions/5595425/what-is-the-best-way-to-compare-floats-for-almost-equality-in-python) to know how to do this

**Reply by toddabraham** [reply]
@AlbertoSinigaglia Thank you for that explanation.  I think I realize now what's happening.  Once I actually deployed the lottery contract, the conversion on rinkeby followed the current price of eth, if I'm reading this result correctly.

![image](https://user-images.githubusercontent.com/26919212/134685546-c6bf1f11-8237-4561-a19c-18f42fba191e.png)

**Reply by AlbertoSinigaglia** [reply]
Sorry, I don't understand what you mean... since the moment you upload the contract, it will fetch the latest conversion form the Chainlink Oracle, and so it can change in time (this is why you have to get also the conversion in the Python test)
